File: Javascript/optimize_list_sum.js
function Sum() {
    this.sums = [0]

    // Time complexity: O(n)
    // Space complexity: O(n)
    this.BuildSum = function (nums) {
        var sum = 0
        for (i in nums) {
            sum += nums[i]
            this.sums.push(sum)
        }
    }

    // Time complexity: O(1)
    this.OptimizeSum = function (start, end) {
        return this.sums[end] - this.sums[start]
    }
}

nums = [1,2,3,4,5,6,7,8,9,10]
var s = new Sum()
s.BuildSum(nums)
console.log(s.OptimizeSum(2,5))

File: Python/optimize_list_sum.py
class Solution():
    def __init__(self, nums):
        self.name = 'S->13'
        self.sums = [0]
        sum = 0
        for n in nums:
            sum += n
            self.sums.append(sum)

    # Time complexity: O(1)
    # Space complexity: O(n)
    def OptimizeListSum(self, start, end):
        return self.sums[end] - self.sums[start]

nums = [1,2,3,4,5,6,7,8,9,10]
s = Solution(nums)
print(s.OptimizeListSum(2,5))
